benchexec/container.py

```diff
@@ -529,33 +542,90 @@ def duplicate_mount_hierarchy(mount_base, temp_base, work_base, dir_modes):
         temp_path = temp_base + mountpoint
 
         if mode == DIR_OVERLAY:
-            overlay_count += 1
-            work_path = work_base + b"/" + str(overlay_count).encode()
-            os.makedirs(temp_path, exist_ok=True)
-            os.makedirs(work_path, exist_ok=True)
-            try:
-                # Previous mount in this place not needed if replaced with overlay dir.
-                libc.umount(mount_path)
-            except OSError as e:
-                logging.debug(e)
-            try:
-                make_overlay_mount(mount_path, mountpoint, temp_path, work_path)
-            except OSError as e:
-                mp = mountpoint.decode()
-                raise OSError(
-                    e.errno,
-                    f"Creating overlay mount for '{mp}' failed: {os.strerror(e.errno)}. "
-                    f"Please use other directory modes, "
-                    f"for example '--read-only-dir {shlex.quote(mp)}'.",
-                )
+            if os.path.ismount(mount_path):
+                try:
+                    # Previous mount in this place not needed if replaced with overlay dir.
+                    libc.umount(mount_path)
+                except OSError as e:
+                    logging.debug(e)
+
+            if use_fuse and fuse_overlay_mount_path:
+                fuse_mount_path = fuse_overlay_mount_path + mountpoint
+                make_bind_mount(fuse_mount_path, mount_path)
+            else:
+                overlay_count += 1
+                os.makedirs(temp_path, exist_ok=True)
+                work_path = work_base + b"/" + str(overlay_count).encode()
+                os.makedirs(work_path, exist_ok=True)
+                try:
+                    make_overlay_mount(mount_path, mountpoint, temp_path, work_path)
+                except OSError as e:
+                    # Resort to fuse-overlayfs if kernel overlayfs is not available.
```

Comment:
Because you asked for advice on refactoring: I think there is no real perfect solution. The best that I could think of is to move everything inside this `except` block into a separate function called `fallback_to_fuse_overlayfs` or similar.

It is not perfect because we modify `fuse_overlay_mount_path` in this block, so the fallback function would need to return that value and we need to call it like `fuse_overlay_mount_path = fallback_to_fuse_overlayfs(...)`. This is a little bit uncommon, but we can document it as "If successful, the function sets up a fuse-overlafys mount and returns the path to it, which can be used for subsequent overlay mounts" or so.

Then we get at least rid of a large chunk of deeply nested code inside the already large function.

Reply:
Hm. Then I don't really have a good idea how to improve this.